File: 94. Binary Tree Inorder Traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        
        ans = []
	    # Recursive function
        def inorder(root):
			# Base case
            if root is None:
                return
            
			
			# Travel left then right
            inorder(root.left)
            # Record node
            ans.append(root.val)
            inorder(root.right)
            
            
        inorder(root)
        return ans

        

File: 94. Binary Tree Inorder Traversal/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> ans;
        stack<TreeNode*> toVisit;

        while(1){
            while(root != NULL){
                toVisit.push(root);
                root = root->left;
            };
            if(toVisit.empty()){
                break;
            }
            root = toVisit.top();
            toVisit.pop();

            ans.push_back(root->val);
            root = root->right;
        }

        return ans;
        
    }
};
